group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -0,0 +1,448 @@
+            // the latest group generation information from the JoinResponse.
+            // So let's return a REBALANCE_IN_PROGRESS to let consumer handle it gracefully.
+            throw Errors.REBALANCE_IN_PROGRESS.exception();
+        }
+    }
+
+    /**
+     * Validates an OffsetCommit request for a consumer group.
+     *
+     * @param group     The consumer group.
+     * @param context   The request context.
+     * @param request   The actual request.
+     */
+    public void validateOffsetCommitForConsumerGroup(
+        ConsumerGroup group,
+        RequestContext context,
+        OffsetCommitRequestData request
+    ) throws KafkaException {
+        if (request.generationIdOrMemberEpoch() < 0 && group.members().isEmpty()) {
+            // When the generation id is -1, the request comes from either the admin client
+            // or a consumer which does not use the group management facility. In this case,
+            // the request can commit offsets if the group is empty.
+            return;
+        }
+
+        if (!group.hasMember(request.memberId())) {
+            throw Errors.UNKNOWN_MEMBER_ID.exception();
+        }
+
+        final int memberEpoch = group.getOrMaybeCreateMember(request.memberId(), false).memberEpoch();
+        if (request.generationIdOrMemberEpoch() != memberEpoch) {
+            // Consumers using the new consumer group protocol (KIP-848) should be using the
+            // OffsetCommit API >= 9. As we don't support upgrading from the old to the new
+            // protocol yet, we return an UNSUPPORTED_VERSION error if an older version is
+            // used. We will revise this when the upgrade path is implemented.
+            if (context.header.apiVersion() >= 9) {
+                throw Errors.STALE_MEMBER_EPOCH.exception();
+            } else {
+                throw Errors.UNSUPPORTED_VERSION.exception();
+            }
+        }
+    }
+
+    /**
+     * Handles an OffsetCommit request.
+     *
+     * @param context The request context.
+     * @param request The OffsetCommit request.
+     *
+     * @return A Result containing the OffsetCommitResponseData response and
+     *         a list of records to update the state machine.
+     */
+    public CoordinatorResult<OffsetCommitResponseData, Record> commitOffset(
+        RequestContext context,
+        OffsetCommitRequestData request
+    ) throws ApiException {
+        validateOffsetCommit(context, request);
+
+        final long currentTimeMs = time.milliseconds();
+        // "default" expiration timestamp is defined as now + retention. The retention may be overridden
+        // in versions from v2 to v4. Otherwise, the retention defined on the broker is used. If an explicit
+        // commit timestamp is provided (v1 only), the expiration timestamp is computed based on that.
+        final OptionalLong expireTimestampMs = request.retentionTimeMs() == OffsetCommitRequest.DEFAULT_RETENTION_TIME ?
```

Comment:
This one comes from [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupCoordinatorAdapter.scala#L383).